### PR TITLE
#9978 Entity getters and setters

### PR DESCRIPTION
**Pull request by burzum**
See 	#9978

### CODE REVIEW

File: src/Datasource/EntityTrait.php [hunk 1]
```diff
@@ -418,6 +418,36 @@ public function hiddenProperties($properties = null)
     }
 
     /**
+     * Sets hidden properties.
+     *
+     * @param null|array $properties Either an array of properties to treat as virtual or null to get properties
+     * @param bool $merge Merge the new properties with the existing. By default false.
+     * @return self
```
Comment (dereuromark) [inline src/Datasource/EntityTrait.php]: Should be $this then now, right?

Comment (burzum) [inline src/Datasource/EntityTrait.php]: Are you sure? The Query builder for example is using self as well.

File: src/Datasource/EntityTrait.php [hunk 2]
```diff
@@ -418,6 +418,36 @@ public function hiddenProperties($properties = null)
...
+     */
+    public function setHiddenProperties(array $properties, $merge = false)
```
Comment (ADmad) [inline src/Datasource/EntityTrait.php]: Can we drop the `Properties` suffix from all new methods?  It makes the names too long, plus it's not consistently used for all methods so doesn't bring much value.

Comment (burzum) [inline src/Datasource/EntityTrait.php]: Sure

File: src/Datasource/EntityTrait.php (hunk 1 above)
Comment (dereuromark) [inline src/Datasource/EntityTrait.php]: I think it follows the recent discussion about self/static or $this, and self is basically a new instance or clean, whereas $this means the chained object instance itself again.

File: src/Datasource/EntityTrait.php [hunk 3]
```diff
@@ -418,6 +418,36 @@ public function hiddenProperties($properties = null)
     }
 
     /**
+     * Sets hidden properties.
+     *
+     * @param null|array $properties Either an array of properties to treat as virtual or null to get properties
```
Comment (markstory) [inline src/Datasource/EntityTrait.php]: There is no option of `null` anymore.

File: src/Datasource/EntityTrait.php [hunk 4]
```diff
@@ -936,18 +1101,49 @@ public function invalid($field = null, $value = null, $overwrite = false)
...
+     * You can also call this method with an array of properties, in which case they
+     * will each take the accessibility value specified in the second argument.
+     *
+     * ### Example:
+     *
+     * ```
+     * $entity->accessible('id', true); // Mark id as not protected
+     * $entity->accessible('author_id', false); // Mark author_id as protected
+     * $entity->accessible(['id', 'user_id'], true); // Mark both properties as accessible
+     * $entity->accessible('*', false); // Mark all properties as protected
```
Comment (markstory) [inline src/Datasource/EntityTrait.php]: Example uses the old method.

File: src/Datasource/EntityTrait.php [hunk 5]
```diff
@@ -965,6 +1161,44 @@ public function accessible($property, $set = null)
...
+            $this->_accessible[$property] :
+            null;
+
+        return ($value === null && !empty($this->_accessible['*'])) || $value;
+    }
+
+    /**
+     * Returns the alias of the repository from which this entity came from.
+     *
+     * @return string|self
```
Comment (markstory) [inline src/Datasource/EntityTrait.php]: Just string now right?

**Comment by burzum** [comment]
@markstory fixed the doc blocks.

File: src/Datasource/EntityTrait.php [hunk 6]
```diff
@@ -418,6 +418,36 @@ public function hiddenProperties($properties = null)
...
+
+        return $this;
+    }
+
+    /**
+     * Gets the hidden properties.
+     *
+     * @return array
+     */
+    public function getHidden()
```
Comment (markstory) [inline src/Datasource/EntityTrait.php]: Should the old methods be marked as deprecated?

File: src/Datasource/EntityTrait.php [hunk 7]
```diff
@@ -681,7 +755,22 @@ public function dirty($property = null, $isDirty = null)
         $this->_dirty[$property] = true;
         unset($this->_errors[$property], $this->_invalid[$property]);
 
-        return true;
+        return $this;
```
Comment (markstory) [inline src/Datasource/EntityTrait.php]: Line 752 doesn't return `$this`. Should it?

File: src/Datasource/EntityTrait.php [hunk 8]
```diff
@@ -729,6 +818,92 @@ public function isNew($new = null)
...
+    }
+
+    /**
+     * Sets error messages to the entity
+     *
+     * ## Example
+     *
+     * ```
+     * // Sets the error messages for multiple fields at once
+     * $entity->errors(['salary' => ['message'], 'name' => ['another message']);
```
Comment (markstory) [inline src/Datasource/EntityTrait.php]: This code example is missing a `]`.

**Comment by lorenzo** [comment]
What about the interface?

**Comment by markstory** [comment]
@lorenzo I don't think we can add new methods to the interface without breaking backwards compatibility. We'd have to add the new methods as documentation only method prototypes.

**Comment by burzum** [comment]
@markstory if we can't make the change to the interface yet, what else is missing so that we can get it merged?

**Comment by chinpei215** [comment]
I think you can use `getError()` and `setErrors()` in the deprecated `errors()` method.

[L948-L953](https://github.com/cakephp/cakephp/blob/b49e6ed4e56dc93fc8fc98bce8ad19f85531ccc1/src/Datasource/EntityTrait.php#L948-L953)
```php
// $errors = isset($this->_errors[$field]) ? $this->_errors[$field] : [];
// if ($errors) {
//     return $errors;
// }
// return $this->_nestedErrors($field);

return $this->getError($field);
```

[L960-L967](https://github.com/cakephp/cakephp/blob/b49e6ed4e56dc93fc8fc98bce8ad19f85531ccc1/src/Datasource/EntityTrait.php#L960-L967)
```php
// foreach ($field as $f => $error) {
//     $this->_errors += [$f => []];
//     $this->_errors[$f] = $overwrite ?
//         (array)$error :
//         array_merge($this->_errors[$f], (array)$error);
// }
// return $this;

return $this->setErrors($field);
```

Seems asymmetric to me though.

**Pull request merged**